extendr-api/src/wrapper/complexes.rs

```diff
@@ -0,0 +1,137 @@
+use super::scalar::{Rcplx, C64};
+use super::*;
+use std::iter::FromIterator;
+
+/// An obscure `NA`-aware wrapper for R's complex vectors.
+/// Can be used to iterate over vectors obtained from R
+/// or to create new vectors that can be returned back to R.
+/// ```
+/// use extendr_api::prelude::*;
+/// test! {
+///     let mut vec = (0..5).map(|i| (i as f64).into()).collect::<Complexes>();
+///     vec.iter_mut().for_each(|v| *v = *v + Rcplx::from(10.0));
+///     assert_eq!(vec.elt(0), Rcplx::from(10.0));
+///     let sum = vec.iter().sum::<Rcplx>();
+///     assert_eq!(sum, Rcplx::from(60.0));
+/// }
+/// ```  
+#[derive(Debug, PartialEq, Clone)]
+pub struct Complexes {
+    pub(crate) robj: Robj,
+}
+
+crate::wrapper::macros::gen_vector_wrapper_impl!(
+    vector_type: Complexes,
+    scalar_type: Rcplx,
+    primitive_type: C64,
+    r_prefix: COMPLEX,
+    SEXP: CPLXSXP,
+    doc_name: complex,
+    altrep_constructor: make_altcomplex_from_iterator,
+);
+
+impl Complexes {
+    /// Get a region of elements from the vector.
+    pub fn get_region(&self, index: usize, dest: &mut [Rcplx]) -> usize {
```

Comment:
It feels like this can be a macro. Vector types other than `Logicals` may benefit from it as well.

Reply:
It was originally a macro, but Doubles and Complexes differ in the methods they support.